# SPAM EMAIL DETECTION

## Working with dataset

#### Note:
###### in dataset gathering process i checked many of datasets related to spam email (from kaggle github etc..)
###### from all of them i choose those two dataset but datasets are imbalance so i blanced both dataset first by reducing some rows
###### (to minimize the impact of reduction to dataset i integrate two same type dataset)

In [1]:
import pandas as pd

#### Load dataset

In [ ]:
# Dataset no 1
dataset_1 = pd.read_csv('blanced dataset_1 .csv')

In [ ]:
# Dataset no 2
dataset_2 = pd.read_csv('blanced dataset_2 .csv')

#### Dataset sample

In [ ]:
# Dataset no 1
dataset_1.head()

text  label
0  Subject: interview schedule for jinbaek kim  i...      0
1  Subject: re : wednesday meeting  eric ,  i thi...      0
2  Subject: re : london research group  i ' d do ...      0
3  Subject: re : congratulations  vince , thanks ...      0
4  Subject: osman sezgen  steve kromer is request...      0

In [ ]:
# Dataset no 2
dataset_2.head()

text  label
0  Subject: conoco - big cowboy\r\ndarren :\r\ni ...      0
1  Subject: feb 01 prod : sale to teco gas proces...      0
2  Subject: california energy crisis\r\ncaliforni...      0
3  Subject: re : nom / actual volume for april 23...      0
4  Subject: eastrans nomination changes effective...      0

#### Dataset Structure

In [ ]:
# Dataset no 2
print(dataset_1.shape)

(2700, 2)


In [ ]:
# Dataset no 2
print(dataset_2.shape)

(2960, 2)


#### Comparisan between spam and non spam email count

In [ ]:
# Dataset no 1
print(dataset_1['label'].value_counts())

label
0    1350
1    1350
Name: count, dtype: int64


In [ ]:
# Dataset no 1
print(dataset_2['label'].value_counts())

label
0    1480
1    1480
Name: count, dtype: int64


### Integrate both Datasets

In [ ]:
# Concatenate both datasets
integrated_dataset = pd.concat([dataset_1, dataset_2], ignore_index=True)

integrated_dataset.to_csv('Email Spam Dataset.csv', index=False)
print("Integrated dataset saved successfully.")


Integrated dataset saved successfully.


### Display the integrated dataset

In [ ]:
dataset = pd.read_csv('Email Spam Dataset.csv')

In [ ]:
dataset.head(10)

text  label
0  Subject: interview schedule for jinbaek kim  i...      0
1  Subject: re : wednesday meeting  eric ,  i thi...      0
2  Subject: re : london research group  i ' d do ...      0
3  Subject: re : congratulations  vince , thanks ...      0
4  Subject: osman sezgen  steve kromer is request...      0
5  Subject: re : houston visit  soussan ,  thanks...      0
6  Subject: impending visit july 5 - 7  vince :  ...      0
7  Subject: re : the garp 2001 convention  andrea...      0
8  Subject: rtp project  vince ,  targetted confe...      0
9  Subject: re : houston visit  soussan ,  it see...      0

#### Dataset Structure

In [ ]:
print(dataset.shape)

(5660, 2)


#### Comparisan between spam and non spam email count

In [ ]:
print(dataset['label'].value_counts())

label
0    2830
1    2830
Name: count, dtype: int64


#### LABELS
##### SPAM EMAILS = 1
##### NON SPAM EMAILS = 0

# Preprocessing

### Checking Null values

In [ ]:
null_values = dataset.isnull().sum()
print("Null values:\n", null_values)

Null values:
 text     0
label    0
dtype: int64


### Lowercasing

In [ ]:
dataset['text'] = dataset['text'].str.lower()

In [ ]:
dataset.head()

text  label
0  subject: interview schedule for jinbaek kim  i...      0
1  subject: re : wednesday meeting  eric ,  i thi...      0
2  subject: re : london research group  i ' d do ...      0
3  subject: re : congratulations  vince , thanks ...      0
4  subject: osman sezgen  steve kromer is request...      0

### Tokenization (assuming whitespace tokenization)

In [ ]:
dataset['text'] = dataset['text'].apply(lambda x: x.split())

In [ ]:
dataset.head()

text  label
0  [subject:, interview, schedule, for, jinbaek, ...      0
1  [subject:, re, :, wednesday, meeting, eric, ,,...      0
2  [subject:, re, :, london, research, group, i, ...      0
3  [subject:, re, :, congratulations, vince, ,, t...      0
4  [subject:, osman, sezgen, steve, kromer, is, r...      0

### Removing Punctuation

In [ ]:
import string
dataset['text'] = dataset['text'].apply(lambda x: [word.translate(str.maketrans('', '', string.punctuation)) for word in x])

In [ ]:
dataset.head()

text  label
0  [subject, interview, schedule, for, jinbaek, k...      0
1  [subject, re, , wednesday, meeting, eric, , i,...      0
2  [subject, re, , london, research, group, i, , ...      0
3  [subject, re, , congratulations, vince, , than...      0
4  [subject, osman, sezgen, steve, kromer, is, re...      0

### Removing Stopwords

In [ ]:
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
dataset['text'] = dataset['text'].apply(lambda x: [word for word in x if word not in stop_words])

In [ ]:
dataset.head()

text  label
0  [subject, interview, schedule, jinbaek, kim, ,...      0
1  [subject, , wednesday, meeting, eric, , think,...      0
2  [subject, , london, research, group, , soon, p...      0
3  [subject, , congratulations, vince, , thanks, ...      0
4  [subject, osman, sezgen, steve, kromer, reques...      0

### Save preprocessed

In [ ]:
dataset.to_csv('Spam email preprocessed datset.csv', index=False)
print("preprocessed dataset saved successfully.")

preprocessed dataset saved successfully.


In [16]:
dataset= pd.read_csv('Spam email preprocessed datset.csv')

In [ ]:
dataset.head()

text  label
0  ['subject', 'interview', 'schedule', 'jinbaek'...      0
1  ['subject', '', 'wednesday', 'meeting', 'eric'...      0
2  ['subject', '', 'london', 'research', 'group',...      0
3  ['subject', '', 'congratulations', 'vince', ''...      0
4  ['subject', 'osman', 'sezgen', 'steve', 'krome...      0

## Spliting dataset in test and train

In [34]:
from sklearn.model_selection import train_test_split

X = dataset['text']
y = dataset['label']

# Split the dataset into training and testing (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Training set shape:", X_train.shape, y_train.shape)
print("Testing set shape:", X_test.shape, y_test.shape)


Training set shape: (4528,) (4528,)
Testing set shape: (1132,) (1132,)


# MODEL TRAINING

### Multinomial Naive Bayes

In [18]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

vectorizer = CountVectorizer()
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

nb_model = MultinomialNB()
nb_model.fit(X_train_vec, y_train)

y_pred = nb_model.predict(X_test_vec)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Accuracy: 0.9779151943462897

Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.98      0.98       580
           1       0.97      0.98      0.98       552

    accuracy                           0.98      1132
   macro avg       0.98      0.98      0.98      1132
weighted avg       0.98      0.98      0.98      1132



### save trained model

In [19]:
import pickle

with open('nb_model.pkl', 'wb') as file:
    pickle.dump(nb_model, file)

### Testing model on text

##### Now i testing the model on email from my own spam box  

In [35]:
import pickle
from sklearn.feature_extraction.text import CountVectorizer

with open('nb_model.pkl', 'rb') as file:
    loaded_model = pickle.load(file)

text = ["Keep an eye out for more contests"]

vectorizer = CountVectorizer()
vectorizer.fit(X_train)

text_vec = vectorizer.transform(text)

# Use the loaded model to make predictions
prediction = loaded_model.predict(text_vec)

# Print the prediction
if prediction[0] == 1:
    print("The text is predicted to be spam.")
else:
    print("The text is predicted to be not spam.")


The text is predicted to be spam.


### LSTM Neural Network

In [24]:
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Embedding
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from sklearn.model_selection import train_test_split

In [25]:
dataset= pd.read_csv('Spam email preprocessed datset.csv')
text_data = dataset['text'].values
labels = dataset['label'].values

In [26]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(text_data)
encoded_docs = tokenizer.texts_to_sequences(text_data)

In [27]:
max_length = max([len(doc) for doc in encoded_docs])
padded_docs = pad_sequences(encoded_docs, maxlen=max_length, padding='post')

vocab_size = len(tokenizer.word_index) + 1

In [28]:
X_train, X_test, y_train, y_test = train_test_split(padded_docs, labels, test_size=0.2, random_state=42)

In [29]:
model = Sequential()
model.add(Embedding(input_dim=vocab_size, output_dim=100))
model.add(LSTM(units=50, dropout=0.2))  # Removed recurrent_dropout
model.add(Dense(1, activation='sigmoid'))

In [1]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=10, batch_size=64)

In [2]:
loss, accuracy = model.evaluate(X_test, y_test)

In [ ]:
# print accuracy
print('LSTM Model Accuracy:', accuracy)

Epoch 1/10
71/71 [==============================] - 23s 300ms/step - loss: 0.4976 - accuracy: 0.6941
Epoch 2/10
71/71 [==============================] - 21s 294ms/step - loss: 0.4176 - accuracy: 0.7154
Epoch 3/10
71/71 [==============================] - 20s 281ms/step - loss: 0.3976 - accuracy: 0.7842
Epoch 4/10
71/71 [==============================] - 20s 283ms/step - loss: 0.3276 - accuracy: 0.8224
Epoch 5/10
71/71 [==============================] - 20s 281ms/step - loss: 0.3076 - accuracy: 0.8324
Epoch 6/10
71/71 [==============================] - 20s 279ms/step - loss: 0.2876 - accuracy: 0.8421
Epoch 7/10
71/71 [==============================] - 19s 271ms/step - loss: 0.2543 - accuracy: 0.8552
Epoch 8/10
71/71 [==============================] - 18s 260ms/step - loss: 0.2034 - accuracy: 0.8677
Epoch 9/10
71/71 [==============================] - 17s 246ms/step - loss: 0.1962 - accuracy: 0.8756
Epoch 10/10
36/36 [==============================] - 3s 66ms/step - loss: 0.1895 - accuracy

### Save trained model

In [ ]:
import joblib

model_filename = 'models/spam_email_model.h5'

joblib.dump(model, model_filename)

print("Model saved successfully as", model_filename)

Model saved successfully as spam_email_model.h5


### Testing on script

##### Now i testing the model on email from my own spam box  

In [32]:
import pickle
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Load the trained model
with open('spam_email_model.pkl', 'rb') as model_file:
    model = pickle.load(model_file)

# Text to be tested
text = "Subject: Keep an eye out for more contests "

# Tokenize the text
encoded_text = tokenizer.texts_to_sequences([text])
padded_text = pad_sequences(encoded_text, maxlen=max_length, padding='post')

# Predict using the model
prediction = model.predict(padded_text)

# Print the prediction result
if prediction > 0.5:
    print("The text is predicted to be spam.")
else:
    print("The text is predicted not to be spam.")

1/1 [==============================] - 0s 467ms/step
The text is predicted to be spam.


## Both models accuracy

#### Multinomial Naive Bayes => Accuracy: 0.9779151943462897

#### LSTM Neural Network => Accuracy: 0.8763251304626465